trin-core/src/utp/packets.rs

```diff
@@ -113,90 +99,82 @@ pub struct PacketHeader {
-            timestamp_microseconds: u32::from_be_bytes([bytes[4], bytes[5], bytes[6], bytes[7]]),
-            timestamp_difference_microseconds: u32::from_be_bytes([
-                bytes[8], bytes[9], bytes[10], bytes[11],
-            ]),
-            wnd_size: u32::from_be_bytes([bytes[12], bytes[13], bytes[14], bytes[15]]),
-            seq_nr: u16::from_be_bytes([bytes[16], bytes[17]]),
-            ack_nr: u16::from_be_bytes([bytes[18], bytes[19]]),
-        }
-    }
-
+    /// Returns the packet's version.
     pub fn get_version(&self) -> u8 {
-        self.type_ver & 0xf
+        self.type_ver & 0x0F
     }
 
-    pub fn set_version(&mut self, int: u8) {
-        self.type_ver = (self.type_ver & 0xf0) | (int & 0xf)
+    /// Sets the type of packet to the specified type.
+    pub fn set_type(&mut self, t: PacketType) {
+        let version = 0x0F & self.type_ver;
+        self.type_ver = u8::from(t) << 4 | version;
     }
 
+    /// Returns the packet's type.
     pub fn get_type(&self) -> PacketType {
         PacketType::try_from(self.type_ver >> 4).unwrap()
     }
+}
 
-    pub fn set_type(&mut self, t: PacketType) {
-        self.type_ver = (self.type_ver & 0xf) | (u8::from(t) << 4)
-    }
-
-    pub fn set_connection_id(&mut self, connection_id: u16) {
-        self.connection_id = connection_id;
-    }
+impl<'a> TryFrom<&'a [u8]> for PacketHeader {
+    // TODO: Refactor this to use anyhow crate
+    type Error = &'static str;
+    /// Reads a byte buffer and returns the corresponding packet header.
+    /// It assumes the fields are in network (big-endian) byte order,
+    /// preserving it.
+    fn try_from(buf: &[u8]) -> Result<Self, Self::Error> {
+        // Check length
+        if buf.len() < HEADER_SIZE {
+            return Err("The packet is too small");
+        }
 
-    pub fn set_timestamp(&mut self, timestamp_microseconds: u32) {
-        self.timestamp_microseconds = timestamp_microseconds;
-    }
+        // Check version
+        if buf[0] & 0x0F != VERSION {
+            return Err("Unsupported packet version");
+        }
 
-    pub fn set_timestamp_difference(&mut self, timestamp_difference_microseconds: u32) {
-        self.timestamp_difference_microseconds = timestamp_difference_microseconds;
-    }
+        // Check packet type
+        if let Err(e) = PacketType::try_from(buf[0] >> 4) {
+            return Err(e);
+        }
 
-    pub fn set_wnd_size(&mut self, wnd_size: u32) {
-        self.wnd_size = wnd_size;
+        Ok(PacketHeader {
+            type_ver: buf[0],
+            extension: buf[1],
+            connection_id: u8_to_unsigned_be!(buf, 2, 3, u16),
+            timestamp_microseconds: u8_to_unsigned_be!(buf, 4, 7, u32),
+            timestamp_difference_microseconds: u8_to_unsigned_be!(buf, 8, 11, u32),
+            wnd_size: u8_to_unsigned_be!(buf, 12, 15, u32),
+            seq_nr: u8_to_unsigned_be!(buf, 16, 17, u16),
+            ack_nr: u8_to_unsigned_be!(buf, 18, 19, u16),
+        })
```

Comment:
Huh, interesting, why remove the `decode()` in favor of in-lining here? I don't really have a preference, but I my first instinct was that having a `PacketHeader::decode` seemed like a reasonable choice.

Reply:
I think we are creating a simple `PacketHeader` object here and I'm not sure if using `decode()` makes sense anymore, because we are not `decoding` anything anymore, we are creating a new type instance.